lib/flex.js

```diff
@@ -0,0 +1,66 @@
+const Configuration = require('./configuration/configuration')
+const extractValidationStats = require('./stats/extractValidationStats')
+
+// Main logic Processor of mergeable
+const executeMergeable = async (context, registry) => {
+  if (registry === undefined) {
+    registry = { validators: new Map(), actions: new Map() }
+  }
+
+  // first fetch the configuration
+  let config = await Configuration.instanceWithContext(context)
+
+  // go through the settings and register all the validators
+  await config.registerValidatorsAndActions(registry)
+
+  // do pre validation actions
+  await processPreActions(context, registry, config)
+
+  config.settings.forEach(rule => {
+    Promise.all(getValiatorPromises(context, registry, rule))
+      .then((result) => {
+        const translatedOutput = extractValidationStats(result)
+
+        Promise.all(getActionPromises(context, registry, rule, translatedOutput))
+      })
+  })
+}
+
+const getValiatorPromises = (context, registry, rule) => {
+  console.log(context)
+  const validateFuncCall = (validator, context, validation) => validator.validate(context, validation)
+
+  return createPromises(rule.validate, 'validators', validateFuncCall, context, registry)
+}
+
+// call all action classes' beforeValidate, regardless of whether they are in failure or pass situation
+const processPreActions = async (context, registry, config) => {
+  registry.actions.forEach(action => {
+    action.beforeValidate({ context })
+  })
+}
+
+const getActionPromises = (context, registry, rule, result) => {
+  const actions = rule[result.validationStatus]
+
+  const afterValidateFuncCall = (actionClass, context, action, result) => actionClass.afterValidate(context, action, result)
+
+  return createPromises(actions, 'actions', afterValidateFuncCall, context, registry, result)
+}
+
+const createPromises = (arrayToIterate, registryName, funcCall, context, registry, result) => {
+  let promises = []
+
+  arrayToIterate.forEach(element => {
+    let key = element.do
+
+    let actionClass = registry[registryName].get(key)
```

Comment:
I think this variable name should probably just be `klass`  since it can be any type of class. and `class` is a reserved word.

Reply:
Or something else more general.

Reply:
`klass` is cool, sold